homeassistant/components/media_player/apple_tv.py

```diff
@@ -71,34 +71,53 @@ def async_setup_platform(hass, config, async_add_entities,
     details = pyatv.AppleTVDevice(name, host, login_id)
     session = async_get_clientsession(hass)
     atv = pyatv.connect_to_apple_tv(details, hass.loop, session=session)
-    entity = AppleTvDevice(atv, name, start_off)
+    entity = AppleTvDevice(hass, atv, name, start_off)
 
-    yield from async_add_entities([entity], update_before_add=True)
+    def on_hass_stop(event):
+        """Stop push updates when hass stops."""
+        atv.push_updater.stop()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
+
+    yield from async_add_entities([entity])
 
 
 class AppleTvDevice(MediaPlayerDevice):
     """Representation of an Apple TV device."""
 
-    def __init__(self, atv, name, is_off):
+    def __init__(self, hass, atv, name, is_off):
         """Initialize the Apple TV device."""
+        self._hass = hass
         self._atv = atv
         self._name = name
         self._is_off = is_off
         self._playing = None
         self._artwork_hash = None
+        self._atv.push_updater.listener = self
+        self._atv.push_updater.start()
 
     @callback
     def _set_power_off(self, is_off):
         self._playing = None
         self._artwork_hash = None
         self._is_off = is_off
+        if is_off:
+            self._atv.push_updater.stop()
+        else:
+            self._atv.push_updater.start()
```

Comment:
Are these methods async friendly ? (`_set_power_off` is marked as an async callback)